src/util/CancellationHandle.h

```diff
@@ -164,30 +173,36 @@ class CancellationHandle {
   /// this will log a warning if this check is not called frequently enough.
   template <typename... ArgTypes>
   AD_ALWAYS_INLINE void throwIfCancelled(
-      const InvocableWithConvertibleReturnType<
+      [[maybe_unused]] const InvocableWithConvertibleReturnType<
           std::string_view, ArgTypes...> auto& detailSupplier,
-      ArgTypes&&... argTypes) {
-    auto state = cancellationState_.load(std::memory_order_relaxed);
-    if (state == CancellationState::NOT_CANCELLED) [[likely]] {
-      return;
-    }
-    if constexpr (WatchDogEnabled) {
-      if (!detail::isCancelled(state)) {
-        pleaseWatchDog(state, detailSupplier, AD_FWD(argTypes)...);
+      [[maybe_unused]] ArgTypes&&... argTypes) {
+    if constexpr (CancellationEnabled) {
+      auto state = cancellationState_.load(std::memory_order_relaxed);
+      if (state == CancellationState::NOT_CANCELLED) [[likely]] {
         return;
       }
+      if constexpr (WatchDogEnabled) {
+        if (!detail::isCancelled(state)) {
+          pleaseWatchDog(state, detailSupplier, AD_FWD(argTypes)...);
+          return;
+        }
+      }
+      throw CancellationException{
+          state, std::invoke(detailSupplier, AD_FWD(argTypes)...)};
     }
-    throw CancellationException{
-        state, std::invoke(detailSupplier, AD_FWD(argTypes)...)};
   }
 
   /// Return true if this cancellation handle has been cancelled, false
   /// otherwise. Note: Make sure to not use this value to set any other atomic
   /// value with relaxed memory ordering, as this may lead to out-of-thin-air
   /// values.
   AD_ALWAYS_INLINE bool isCancelled() const {
-    return detail::isCancelled(
-        cancellationState_.load(std::memory_order_relaxed));
+    if constexpr (CancellationEnabled) {
+      return detail::isCancelled(
+          cancellationState_.load(std::memory_order_relaxed));
+    } else {
+      return false;
+    }
```

Comment:
This should also please the watchdog, and if you do so, 
then you can use it as the impl of the `throwIfCancelled` for everything but the trhwoing.

Reply:
(The only thing is that this currently gets no details for the watchdog pleasing, but you'll figure something out.

Reply:
Hmm I'm not a fan of this member function in the first place. The only place where it's being used is to ensure within openmp there isn't any work started after any cancellation might have happened. It doesn't actually cancel the operation. It just calls throwIfCancelled again after the parallel block.
But I guess I could just document that.

Reply:
Yes, you can leave it as is.
I mean you can argue that if you do this "all blocks are cancelled, but the currently active blocks are run to completion" scheme, then a passed watchdog deadline means that the blocks are probably too large which is in itself also a useful information, so just leave it, thanks for the discussion.